xcm/xcm-builder/src/location_conversion.rs

```diff
@@ -345,6 +345,42 @@ impl<Network: Get<Option<NetworkId>>, AccountId: From<[u8; 20]> + Into<[u8; 20]>
 	}
 }
 
+/// Converts a location which is a top-level relay chain (which provides its own consensus) into a 32-byte `AccountId`.
+///
+/// This will always result in the *same account ID* being returned for the same Relay-chain, regardless of the relative security of
+/// this Relay-chain compared to the local chain.
+///
+/// Note: No distinction is made between the cases when the given `UniversalLocation` lies within
+/// the same consensus system (i.e. is itself or a parent) and when it is a foreign consensus
+/// system.
+pub struct GlobalConsensusConvertsFor<UniversalLocation, AccountId>(
+	PhantomData<(UniversalLocation, AccountId)>,
+);
+impl<UniversalLocation: Get<InteriorMultiLocation>, AccountId: From<[u8; 32]> + Clone>
+	ConvertLocation<AccountId> for GlobalConsensusConvertsFor<UniversalLocation, AccountId>
+{
+	fn convert_location(location: &MultiLocation) -> Option<AccountId> {
+		let universal_source = UniversalLocation::get();
```

Comment:
Perhaps worth asserting here as well is to make sure that there is only 1 junction here and it is the GlobalConsensus junction.

Reply:
Ideally, this should be a compile-time check, but I don't think it's possible yet without making a substantial change to the underlying MultiLocation enum.

Reply:
"only 1 junction" will not work here, because `UniversalLocation` will be usually (at least in our cases) parachain, so e.g.:
```
pub UniversalLocation: InteriorMultiLocation =
		X2(GlobalConsensus(RelayNetwork::get()), Parachain(ParachainInfo::parachain_id().into()));
```

and also `fn ensure_is_remote` has a check for required `GlobalConsensus` [here](https://github.com/paritytech/polkadot/blob/master/xcm/xcm-builder/src/universal_exports.rs#L36-L39):
```
        let local_net = match universal_local.global_consensus() {
		Ok(x) => x,
		Err(_) => return Err(dest),
	};
```

Reply:
I don't get it, you're making a match so that the remote location is only `Here` below, and I'm saying that you should instead make the check here and ensure that the location is only 1 length long and it's the GlobalConsensus junction.

Reply:
@KiChjang 
this call, makes sure that `location` is remote location (from different global consensus) from the `UniversalLocation`'s point of view,
if `location` does not contain `GlobalConsensus`, this will fail
```
let (remote_network, remote_location) = ensure_is_remote(UniversalLocation, location).ok()?;
```
so this is a first check for "remote location from the UniversalLocation's point of view",
so after this call we know that `location` is from different `GlobalConsensus(remote_network)`,

and after this, there is second check for `remote_location`,
and we are strictly interested here just in `MultiLocation { parents: 2, X1(GlobalConsensus(remote_network)) }`,
so `ensure_is_remote` results in:
```
let universal_location = X2(GlobalConsensus(Kusama), Parachain(1000));
let location = MultiLocation { parents: 2, X1(GlobalConsensus(Polkadot)) };
let (remote_network, remote_location) = ensure_is_remote(universal_location, location).ok()?;

assert_eq!(remote_network, Polkadot);
assert_eq!(remote_location, Here);
```

so after `ensure_is_remote` the `remote_location` part if `Here`,
thats why second check is matching strictly only for `Here`

---

or what do you suggest? skip `ensure_is_remote` and just match `location` like this: ?
```
match location {
			MultiLocation { parents, X1(GlobalConsensus(remote_network)) if parents>= 2 => Some(AccountId::from(Self::from_params(&remote_network))),
			_ => None,
		}
```

I think we cannot do this, because of discussion for previous  `GlobalConsensusParachainConvertsFor` here https://github.com/paritytech/polkadot/pull/7016#discussion_r1194286757

---

I mean this `GlobalConsensusConvertsFor` is exactly the same implementation logic as previous `GlobalConsensusParachainConvertsFor`,
both do the "remote location" check at first, 
and the only difference is the second check:
`GlobalConsensusConvertsFor` - matches for strict `Here`
`GlobalConsensusParachainConvertsFor` - matches for strict `X1(Parachain(para_id))`

Reply:
Okay, since the latter is more rigorous, let's just keep it as is.